common/test/test_migrations/utils.py

```diff
@@ -0,0 +1,65 @@
+"""
+General class for test migrations.Based off an implementation provided at
+https://www.caktusgroup.com/blog/2016/02/02/writing-unit-tests-django-migrations/
+"""
+# pylint: disable=redefined-outer-name
+
+from django.db import connection
+from django.db.migrations.executor import MigrationExecutor
+from django.db.migrations.recorder import MigrationRecorder
+from django.test import TransactionTestCase
+from django.core.management import call_command
+
+
+class TestMigrations(TransactionTestCase):
+    """ Base class for testing migrations. """
+    migrate_from = None
+    migrate_to = None
+    app = None
+
+    def setUp(self):
+        super(TestMigrations, self).setUp()
+        assert self.migrate_from and self.migrate_to, \
+            "TestCase '{}' must define migrate_from and migrate_from properties".format(type(self).__name__)
+        assert self.app, "app must be define in the TestCase"
+        self.initial_migration = MigrationRecorder.Migration.objects.filter(app=self.app).last().name
+        self.executor = MigrationExecutor(connection)
+
+    def tearDown(self):
+        """ Back to initial migration """
+        super(TestMigrations, self).tearDown()
+        call_command("migrate", self.app, self.initial_migration)
+        self._check_migration_state(self.initial_migration)
+
+    def _check_migration_state(self, migration_name):
+        """ Veirfy the migration from djano migration table"""
+        migration_state = MigrationRecorder.Migration.objects.filter(app=self.app).last()
+        self.assertEqual(migration_state.name, migration_name)
+
+    def execute_migration(self, migrate_from, migrate_to):
+        """
+        Execute migration from state to another.
+        """
+        # Reverse to the original migration
+        self.executor.migrate([(self.app, self.migrate_from)])
```

Comment:
Here's what I don't understand. Say a model has three migrations: `0001`, `0002`, and `0003`. We want to test that the migration from `0002` to `0003` goes smoothly. At the beginning of this test case, we'll be at `0003`. At this line, would we be rolling back our migration, or trying to apply `0002` on top of `0003`? Isn't this the problem you had with `migrate_backwards`, that caused you to switch to `call_command`? Why don't all these cases use `call_command` instead of `self.executor.migrate`, especially since we'll be executing these migrations _with_ that command?